types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,630 @@
+	wl_list_for_each(tmp_conn, &device->connectors, link) {
+		if (tmp_conn->output == output) {
+			connector = tmp_conn;
+			break;
+		}
+	}
+
+	if (!connector) {
+		wlr_log(WLR_DEBUG, "Output has not been offered");
+		return;
+	}
+
+	if (connector->active_lease) {
+		wlr_drm_lease_v1_revoke(connector->active_lease);
+	}
+
+	struct wl_resource *resource, *tmp;
+	wl_resource_for_each_safe(resource, tmp, &connector->resources) {
+		wp_drm_lease_connector_v1_send_withdrawn(resource);
+		wl_resource_set_user_data(resource, NULL);
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_list_init(wl_resource_get_link(resource));
+	}
+
+	wl_resource_for_each(resource, &device->requests) {
+		struct wlr_drm_lease_request_v1 *request =
+			wlr_drm_lease_request_v1_from_resource(resource);
+		request->invalid = true;
+	}
+
+	wl_list_remove(&connector->link);
+	wl_list_init(&connector->link);
+	free(connector);
+}
+
+static void wlr_drm_lease_device_v1_destroy(
+		struct wlr_drm_lease_device_v1 *device) {
+	struct wlr_drm_backend *backend =
+			get_drm_backend_from_backend(device->backend);
+	wlr_log(WLR_DEBUG, "Destroying wlr_drm_lease_device_v1 for %s",
+			backend->name);
+
+	struct wl_resource *resource;
+	struct wl_resource *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &device->resources) {
+		wl_resource_destroy(resource);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->requests) {
+		wl_resource_destroy(resource);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->leases) {
+		struct wlr_drm_lease_v1 *lease =
+			wlr_drm_lease_v1_from_resource(resource);
+		wlr_drm_lease_v1_revoke(lease);
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector, *tmp;
+	wl_list_for_each_safe(connector, tmp, &device->connectors, link) {
+		wl_list_remove(&connector->link);
+		wl_list_init(&connector->link);
+		free(connector);
+	}
+
+	free(device);
+}
+
+static void handle_backend_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_drm_lease_device_v1 *device =
+		wl_container_of(listener, device, backend_destroy);
+
+	wlr_drm_lease_device_v1_destroy(device);
+}
+
+static struct wlr_drm_lease_device_v1 *wlr_drm_lease_device_v1_create(
+		struct wl_display *display, struct wlr_backend *backend) {
+	assert(display && backend && wlr_backend_is_drm(backend));
+
+	struct wlr_drm_backend *drm_backend =
+			get_drm_backend_from_backend(backend);
+	wlr_log(WLR_DEBUG, "Creating wlr_drm_lease_device_v1 for %s",
+			drm_backend->name);
+
+	struct wlr_drm_lease_device_v1 *lease_device =
+		calloc(1, sizeof(struct wlr_drm_lease_device_v1));
+
+	if (!lease_device) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_device_v1");
+		return NULL;
+	}
+
+	lease_device->backend = backend;
+	wl_list_init(&lease_device->resources);
+	wl_list_init(&lease_device->connectors);
+	wl_list_init(&lease_device->requests);
+	wl_list_init(&lease_device->leases);
+	wl_list_init(&lease_device->link);
```

Comment:
We should insert the device into the manager's list here, instead of just calling `wl_list_init`.

Reply:
I was inserting the device into the manager's list outside of the function, will move it inside.